engines/cufile.c

```diff
@@ -0,0 +1,586 @@
+ *
+ */
+
+#include <stdlib.h>
+#include <unistd.h>
+#include <errno.h>
+#include <string.h>
+#include <sys/time.h>
+#include <sys/resource.h>
+#include <cufile.h>
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <pthread.h>
+
+#include "../fio.h"
+#include "../lib/pow2.h"
+#include "../optgroup.h"
+#include "../lib/memalign.h"
+
+#define ALIGNED_4KB(v) (((v) & 0x0fff) == 0)
+
+#define LOGGED_BUFLEN_NOT_ALIGNED     0x01
+#define LOGGED_OFFSET_NOT_ALIGNED     0x02
+#define LOGGED_GPU_OFFSET_NOT_ALIGNED 0x03
+#define GPU_ID_SEP ":"
+
+struct cufile_options {
+	struct thread_data *td;
+	char               *gpu_ids;       /* colon-separated list of GPU ids,
+										  one per job */
+	void               *cu_mem_ptr;    /* GPU memory */
+	void               *junk_buf;      /* buffer to simulate cudaMemcpy with
+										  posix I/O write */
+	int                 my_gpu_id;     /* GPU id to use for this job */
+	unsigned int        use_posixio;   /* Use POSIX I/O instead of cuFile API */
+	size_t              total_mem;     /* size for cu_mem_ptr and junk_buf */
+	int                 logged;        /* bitmask of log messages that have
+										  been output, prevent flood */
+};
+
+struct fio_cufile_data {
+	CUfileDescr_t  cf_descr;
+	CUfileHandle_t cf_handle;
+};
+
+struct fio_option options[] = {
+	{
+		.name	= "cufile_gpu_ids",
+		.lname	= "cufile gpu ids",
+		.type	= FIO_OPT_STR_STORE,
+		.off1	= offsetof(struct cufile_options, gpu_ids),
+		.help	= "GPU IDs, one per job, separated by " GPU_ID_SEP,
+		.category = FIO_OPT_C_ENGINE,
+		.group	= FIO_OPT_G_CUFILE,
+	},
+	{
+		.name	= "cufile_posixio",
+		.lname	= "cufile posixio",
+		.type	= FIO_OPT_BOOL,
+		.off1	= offsetof(struct cufile_options, use_posixio),
+		.help	= "Use POSIX I/O instead of cuFile API",
+		.def    = "0",
+		.category = FIO_OPT_C_ENGINE,
+		.group	= FIO_OPT_G_CUFILE,
+	},
+	{
+		.name	= NULL,
+	},
+};
+
+static int running = 0;
+static int cufile_initialized = 0;
+static pthread_mutex_t running_lock = PTHREAD_MUTEX_INITIALIZER;
+
+#define check_cudaruntimecall(fn, rc)								\
+	do {															\
+		cudaError_t res = fn;										\
+		if (res != cudaSuccess) {									\
+			const char *str = cudaGetErrorName(res);				\
+			log_err("cuda runtime api call failed %s:%d : %s\n",	\
+					#fn, __LINE__, str);							\
+			rc = -1;												\
+		} else {													\
+			rc = 0;													\
+		}															\
+	} while(0)
+
+static const char* cuFileGetErrorString(CUfileError_t st)
+{
+	if (IS_CUFILE_ERR(st.err)) {
+		return cufileop_status_error(st.err);
+	}
+	return "unknown";
+}
+
+/**
+ * fio_cufile_init called in job process after fork() or pthread_create().
+ */
+static int fio_cufile_init(struct thread_data *td)
+{
+	struct cufile_options *o = td->eo;
+	char* cur = NULL;
+	char* pos = NULL;
+	CUfileError_t status;
+	int gpu_id = 0;
+	int i;
+	int initialized;
+	int rc;
+
+	pthread_mutex_lock(&running_lock);
+	if (running == 0) {
```

Comment:
If this is running in a separate process to other versions of this engine will that be an issue?

Reply:
Sorry, I'm not sure that I understand the question. This is designed to operate in both multi-process and multi-threaded mode. If you are referring to client/server operation, there is no explicit data transfer implemented by this engine. Everything happens through the filesystem. So, I believe the answer is: "No, there should not be any issue."